privacy-profile-composer/pkg/envoyfilter/filter.go

```diff
@@ -0,0 +1,260 @@
+	// run PII Analysis + OPA directly
+	if f.sidecarDirection == common.Inbound {
+		processBody = true
+	}
+
+	//  If it is an outbound sidecar, then check if it's a request to a third party
+	//  and only process the body in this case
+	if f.sidecarDirection == common.Outbound {
+		thirdPartyURL, err := f.checkIfRequestToThirdParty()
+		if err != nil {
+			log.Println(err)
+			return api.Continue
+		} else if thirdPartyURL == "" {
+			log.Printf("outbound sidecar processed a request to another sidecar in the mesh" +
+				"Prose will process it through the inbound decode function\n")
+			return api.Continue
+		}
+		processBody = true
+	}
+
+	if processBody {
+		return f.processBody(buffer, true)
+	}
+
+	return api.Continue
+}
+
+func (f *Filter) DecodeTrailers(trailers api.RequestTrailerMap) api.StatusType {
+	log.Println(">>> DECODE TRAILERS")
+	log.Printf("%+v", trailers)
+	return api.Continue
+}
+
+func (f *Filter) EncodeHeaders(header api.ResponseHeaderMap, endStream bool) api.StatusType {
+	log.Println("<<< ENCODE HEADERS")
+
+	common.LogEncodeHeaderData(header)
+
+	span := f.tracer.StartSpan("test span in encode headers", zipkin.Parent(f.parentSpanContext))
+	defer span.Finish()
+
+	return api.Continue
+}
+
+// Callbacks which are called in response path
+func (f *Filter) EncodeData(buffer api.BufferInstance, endStream bool) api.StatusType {
+	log.Println("<<< ENCODE DATA")
+	log.Println("  <<About to forward", buffer.Len(), "bytes of data to client>>")
+
+	// if outbound then indirect purpose of use violation
+	// TODO: This is usually data obtained from another service
+	//  but it could also be data obtained from a third party. I.e. a kind of join violation.
+	//  Not sure if we'll run into those cases in the examples we look at.
+	if f.sidecarDirection == common.Outbound {
+		return f.processBody(buffer, false)
+	}
+
+	// if inbound then ignore
+	// we will just address them in the inbound call to the caller svc
+	return api.Continue
+}
+
+func (f *Filter) EncodeTrailers(trailers api.ResponseTrailerMap) api.StatusType {
+	log.Println("<<< ENCODE TRAILERS")
+	log.Printf("%+v", trailers)
+	return api.Continue
+}
+
+func (f *Filter) OnDestroy(reason api.DestroyReason) {
+	f.tracer.Close()
+	f.opa.Stop(context.Background())
+}
+
+func (f *Filter) processBody(buffer api.BufferInstance, isDecode bool) api.StatusType {
+	jsonBody, err := common.GetJSONBody(f.headerMetadata, buffer)
+	if err != nil {
+		log.Println(err)
+		return api.Continue
+	}
+
+	err = f.runPresidioAndOPA(jsonBody, isDecode)
+	if err != nil {
+		log.Println(err)
+		return api.Continue
+	}
+	return api.Continue
+}
+
+func (f *Filter) runPresidioAndOPA(jsonBody []byte, isDecode bool) error {
+	var decodeOrEncode string
+	if isDecode {
+		decodeOrEncode = "decode"
+	} else {
+		decodeOrEncode = "encode"
+	}
+
+	span := f.tracer.StartSpan(fmt.Sprintf("test span in %s body (in runPresidioAndOPA)", decodeOrEncode), zipkin.Parent(f.parentSpanContext))
```

Comment:
at the very least, we shouldn't be putting top level context as a parent of this span. i hope that since we already created span at a higher level, we no longer need to specify parent (as zipkin will automagically propagate it). If that's not the case, then we have to set span we created in encode/decode function as a parent of this span. Also, given we would attach this span to a parent span that already has info about whether it is decoding or encoding, we don't need to include `decodeOrEncode` bit into this span name, which might simplify this function for us a bit.

```suggestion
	span := f.tracer.StartSpan(fmt.Sprintf("test span in %s body (in runPresidioAndOPA)", decodeOrEncode))
```

Reply:
this doesn't apply since decode headers and decode body happen one after another.

Reply:
Propagated parent spans down to all functions via context and collected prose tags up to the top level.